### DISCUSSION TITLE
Slicing capablities

### DISCUSSION THREAD

**Discussion by jamjad333** [srsran/srsRAN_Project#642] [General Help] [answered]
Hi,
Is there anyone who has implemented slicing with srsRAN project by multiple sst or APN's and how RAN slicing work in this project because there is nothing much explained in released notes regarding that.
Any suggestion will be appreciated.thanks!

**Comment by dvdgrgrtt** [comment, marked as answer]
srsRAN provides support for network slicing, see e.g. https://github.com/srsran/srsRAN_Project/blob/main/configs/slicing.yml
You have to configure the UE to transmit on one slice, configure the core to support slices and also tell the core which slice is UE will be using.

One thing to pay attention to (at least with Open5GS, not sure with other cores) is that it interprets the SD has an hexadecimal number so, for instance, you have to set `SD: 42` in the Open5GS configuration and `SD: 0x42` in the srsRAN configuration.

At the moment, we don't support RAN slicing.

**Reply by adithya-anand-999** [reply]
Hi 

I have a similar question and have modified the gNB config file like so
![image](https://github.com/user-attachments/assets/a4d339f9-29b9-40db-9859-7bf10dd391be)

But am receiving the following error
![image](https://github.com/user-attachments/assets/bebd77bf-d27e-4702-8eec-ed142e1b9a36)


Do you see anything wrong with how the gNB config file is set up, has something changed with how you make slices in the config file with the update to the srsRAN_Project.

**Reply by dvdgrgrtt** [reply]
Slicing is now a sub-field of `cell_cfg`:
```yaml
cell_cfg:
  slicing:
    - sst: 1
      sd: 0x01
    - sst: 2
      sd: 0x02
  dl_arfcn: ...
```
We forgot to update the documentation, sorry for that. We'll fix it ASAP.

**Reply by adithya-anand-999** [reply]
Yes I just caught that myself. 

Do you know of any changes with the srsue's when it comes to slicing done on the gNB

**Reply by dvdgrgrtt** [reply]
srsUE doesn't support slicing (we're not developing it anymore and it has very limited 5G features).

**Reply by adithya-anand-999** [reply]
but I have seen people able to connect srsue's to slices configured in the gNB. So setting up 2 slices in the gNB config, then connecting srsue's to a respective slice.

**Reply by dvdgrgrtt** [reply]
I had no idea that could be done, I guess it's an ad-hoc modification for some specific test - maybe @pgawlowicz can help

**Reply by pgawlowicz** [reply]
Please apply the patch from this [thread](https://github.com/srsran/srsRAN_Project/discussions/875). 

In Open5gs config, you need to change:
```
amf:
...
  plmn_support:
    - plmn_id:
        mcc: 001
        mnc: 01
      s_nssai:
        - sst: 1
        - sst: 1
          sd: 000001
        - sst: 2
          sd: 000002
```

Start Open5gc, and edit subscriber data using a browser:
![image](https://github.com/user-attachments/assets/c42f62c0-5051-42ea-b9ec-e1b8428892f9)

Then please use those two configs:
- [gnb_zmq_slicing.yaml.txt](https://github.com/user-attachments/files/17622275/gnb_zmq_slicing.yaml.txt)
- [ue_zmq_slicing.yaml.txt](https://github.com/user-attachments/files/17622274/ue_zmq_slicing.yaml.txt)

The srsUE should connect to slice with S-NSSAI: sst = 2, sd = 000002

![image](https://github.com/user-attachments/assets/194f614e-5d89-452c-a3de-b5b8b7457dcc)

**Reply by adithya-anand-999** [reply]
Thanks for the reply 

I am getting a similar issue to what was found on the issue tagged in the reply. I have a system with 3 srsue's connecting to the srsRAN gNB using GNU Radio Companion. The core network is the docker Open5GS provided in the srsRAN_Project repo. The core and gNB conect, the error occurs after I execute the flowgraph. The following are some output I can see. 

I did modify the code on the srsue side found in the thread tagged in the reply. 

The following are the outputs after executing the flow graph. I have 2 srsue's on slice 000001 and 1 srsue on slice 000002

core network output
![image](https://github.com/user-attachments/assets/1ce644af-4cb4-4b3e-b5c3-bb2c54820fe1)

srsue outputs (they are all identical) 
![image](https://github.com/user-attachments/assets/cd21768b-e099-4187-9d33-c2cc4da8076d)

gNB has no ouptut. 

srsue setup on Open5GS Web GUI
![image](https://github.com/user-attachments/assets/b785209f-5e2c-4e50-90da-7a0fbff55e3c)

srsue config file 
![image](https://github.com/user-attachments/assets/aaf21f79-b4cc-4dd7-b984-9521bfd9b852)

config for gNB is identical outside of modifications for allowed multiple srsue's, those changes are the same as whats found on specs. 

I can share specific log files if required also.

**Reply by pgawlowicz** [reply]
could you share the ngap_pcap file?

**Reply by adithya-anand-999** [reply]
Yes of course 
[gnb_ngap.zip](https://github.com/user-attachments/files/17651465/gnb_ngap.zip)

**Reply by pgawlowicz** [reply]
Could you share your gnb and srsUE configs? 
In case of the dockerized Open5gs you have to modify this [config file](https://github.com/srsran/srsRAN_Project/blob/main/docker/open5gs/open5gs-5gc.yml) and then rebuild the 5gs docker image.

**Reply by adithya-anand-999** [reply]
Yes I did modify that config file and added the changes you mentioned. When it comes to rebuilding I run
sudo docker compose up --build 5gc
after modifying, which I believe rebuilds the docker image. 

gNB config file 
![image](https://github.com/user-attachments/assets/4a16f48b-4127-44ac-afb1-199b16706a9e)


1 srsue config file,  they are relatively identical. I should mention the system without slicing works as expected. 
![image](https://github.com/user-attachments/assets/88dd9f1b-9138-428f-8e56-8afa5d74be80)

**Reply by pgawlowicz** [reply]
hmm, looks good.
Could you try the configs that I shared with only gnb and one srsUE (without the broker)?

**Reply by pgawlowicz** [reply]
Just to be sure. Did you apply this patch? https://github.com/srsran/srsRAN_4G/pull/1214/files

**Reply by adithya-anand-999** [reply]
yes I did apply that patch, just commented out those lines of code. 

Oh wait, I just thought about it. I would have to then rebuild and make the srsRAN_4G project after modifying the code, correct.

**Reply by adithya-anand-999** [reply]
> Just to be sure. Did you apply this patch? https://github.com/srsran/srsRAN_4G/pull/1214/files

Thanks so much, It ended up being that silly mistake lol.

**Comment by AlejandraOliver** [comment]
Hi everyone!

I follow the steps @pgawlowicz mentioned but I have this issue in the open5gs when the gbn tries to connect:
```
open5gs_5gc  | 03/12 20:10:27.334: [amf] INFO: [Added] Number of AMF-UEs is now 1 (../src/amf/context.c:1616)

open5gs_5gc  | 03/12 20:10:27.334: [gmm] INFO: Registration request (../src/amf/gmm-sm.c:1165)

open5gs_5gc  | 03/12 20:10:27.334: [gmm] INFO: [suci-0-001-01-0000-0-0-0123456780]    SUCI (../src/amf/gmm-handler.c:166)

open5gs_5gc  | 03/12 20:10:27.741: [gmm] INFO: [imsi-001010123456780] Registration complete (../src/amf/gmm-sm.c:2146)

open5gs_5gc  | 03/12 20:10:27.741: [amf] INFO: [imsi-001010123456780] Configuration update command (../src/amf/nas-path.c:612)

open5gs_5gc  | 03/12 20:10:27.741: [gmm] INFO:     UTC [2025-03-12T19:10:27] Timezone[0]/DST[0] (../src/amf/gmm-build.c:559)

open5gs_5gc  | 03/12 20:10:27.741: [gmm] INFO:     LOCAL [2025-03-12T20:10:27] Timezone[3600]/DST[0] (../src/amf/gmm-build.c:564)

open5gs_5gc  | 03/12 20:10:27.742: [amf] INFO: [Added] Number of AMF-Sessions is now 1 (../src/amf/context.c:2571)

open5gs_5gc  | 03/12 20:10:27.742: [gmm] WARNING: [imsi-001010123456780] DNN Not Supported OR Not Subscribed in the Slice (../src/amf/gmm-handler.c:1228)

open5gs_5gc  | 03/12 20:10:27.797: [gmm] INFO: [imsi-001010123456780] No GUTI allocated (../src/amf/gmm-sm.c:1443)

open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO: UE Context Release [Action:2] (../src/amf/ngap-handler.c:1698)

open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[1] (../src/amf/ngap-handler.c:1699)

open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO:     SUCI[suci-0-001-01-0000-0-0-0123456780] (../src/amf/ngap-handler.c:1702)

open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO: [Removed] Number of gNB-UEs is now 0 (../src/amf/context.c:2557)
```

Below I show you the configuration I have:
[config (2).zip](https://github.com/user-attachments/files/19215861/config.2.zip)
![image](https://github.com/user-attachments/assets/13daf903-8636-4d78-a82d-7856a3a43055)
![image](https://github.com/user-attachments/assets/18fe2cb3-b852-4f8b-b37d-a2f79d8feeb1)


I want to have 2 UEs, each of them in one slice. Maybe the problem is because open5gs-5gc.yml is using add_users.py and this script is only configured with one UE? I dont know...

(without slicing the network works perfectly, srsrRAN Proyect lastest version).

If you could help me!! Thanks!!

**Reply by rcbarke** [reply]
> Hi everyone!
> 
> I follow the steps @pgawlowicz mentioned but I have this issue in the open5gs when the gbn tries to connect:
> 
> ```
> open5gs_5gc  | 03/12 20:10:27.334: [amf] INFO: [Added] Number of AMF-UEs is now 1 (../src/amf/context.c:1616)
> 
> open5gs_5gc  | 03/12 20:10:27.334: [gmm] INFO: Registration request (../src/amf/gmm-sm.c:1165)
> 
> open5gs_5gc  | 03/12 20:10:27.334: [gmm] INFO: [suci-0-001-01-0000-0-0-0123456780]    SUCI (../src/amf/gmm-handler.c:166)
> 
> open5gs_5gc  | 03/12 20:10:27.741: [gmm] INFO: [imsi-001010123456780] Registration complete (../src/amf/gmm-sm.c:2146)
> 
> open5gs_5gc  | 03/12 20:10:27.741: [amf] INFO: [imsi-001010123456780] Configuration update command (../src/amf/nas-path.c:612)
> 
> open5gs_5gc  | 03/12 20:10:27.741: [gmm] INFO:     UTC [2025-03-12T19:10:27] Timezone[0]/DST[0] (../src/amf/gmm-build.c:559)
> 
> open5gs_5gc  | 03/12 20:10:27.741: [gmm] INFO:     LOCAL [2025-03-12T20:10:27] Timezone[3600]/DST[0] (../src/amf/gmm-build.c:564)
> 
> open5gs_5gc  | 03/12 20:10:27.742: [amf] INFO: [Added] Number of AMF-Sessions is now 1 (../src/amf/context.c:2571)
> 
> open5gs_5gc  | 03/12 20:10:27.742: [gmm] WARNING: [imsi-001010123456780] DNN Not Supported OR Not Subscribed in the Slice (../src/amf/gmm-handler.c:1228)
> 
> open5gs_5gc  | 03/12 20:10:27.797: [gmm] INFO: [imsi-001010123456780] No GUTI allocated (../src/amf/gmm-sm.c:1443)
> 
> open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO: UE Context Release [Action:2] (../src/amf/ngap-handler.c:1698)
> 
> open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[1] (../src/amf/ngap-handler.c:1699)
> 
> open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO:     SUCI[suci-0-001-01-0000-0-0-0123456780] (../src/amf/ngap-handler.c:1702)
> 
> open5gs_5gc  | 03/12 20:10:31.269: [amf] INFO: [Removed] Number of gNB-UEs is now 0 (../src/amf/context.c:2557)
> ```
> 
> Below I show you the configuration I have: [config (2).zip](https://github.com/user-attachments/files/19215861/config.2.zip) ![image](https://private-user-images.githubusercontent.com/114316371/422054393-13daf903-8636-4d78-a82d-7856a3a43055.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4MTU2NTksIm5iZiI6MTc0MTgxNTM1OSwicGF0aCI6Ii8xMTQzMTYzNzEvNDIyMDU0MzkzLTEzZGFmOTAzLTg2MzYtNGQ3OC1hODJkLTc4NTZhM2E0MzA1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxMlQyMTM1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDI5OGM0YWEyY2Y3YzBjZDVhZmU5MTc0YmYxOGNmYzkzMDQyZjkxODAxZmQ4ZjEyMjI4OTVmZjFkOWU4MDBjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2DQ1j-NCq3JrtWLpMYyAQAaMCvQfv8gVoFSYMwhUWaA) ![image](https://private-user-images.githubusercontent.com/114316371/422054577-18fe2cb3-b852-4f8b-b37d-a2f79d8feeb1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4MTU2NTksIm5iZiI6MTc0MTgxNTM1OSwicGF0aCI6Ii8xMTQzMTYzNzEvNDIyMDU0NTc3LTE4ZmUyY2IzLWI4NTItNGY4Yi1iMzdkLWEyZjc5ZDhmZWViMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxMlQyMTM1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYjljZGQ3MzZkNTRjODQ0YjFlM2Y5YTc1MWY1NDY4Y2QxMjU2ODQ3N2YwNDQ4ZWJmYTM0Mjk0NzRlMjRhOWIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.STAEcemVG4WSBM3Yx8qdrPhQqxj9Sjwj9-7qMVwFpxg)
> 
> I want to have 2 UEs, each of them in one slice. Maybe the problem is because open5gs-5gc.yml is using add_users.py and this script is only configured with one UE? I dont know...
> 
> (without slicing the network works perfectly, srsrRAN Proyect lastest version).
> 
> If you could help me!! Thanks!!

Hi, I am having trouble recalling the specific thread this was discussed within, but @pgawlowicz previously worked through this issue with another grad student in our lab. It occurs because this [python script](https://github.com/srsran/srsRAN_Project/blob/main/docker/open5gs/add_users.py) within [srsRAN Project's dockerized version of Open5GS](https://github.com/srsran/srsRAN_Project/tree/main/docker/open5gs) contains the slice service type (sst), but is missing the slice differentiator (sd). 

You need to add a "sd" key to this JSON at the same level as the existing "sst" key, and set it according the value your UE(s) are using.

[Here](https://www.sharetechnote.com/html/5G/5G_NetworkSlicing.html) is a good bit of reading on the foundational 5G SA core network architecture which provides insight into the error. During 5G SA UE registration, the NSSF, UDM, and AMF network functions work together to validate and register the network slice reported by the UE. What you're looking at in this file is the list of all potential slices that each UE's request is validated against.

Piotr can keep me honest, though I'm not sure there's an easy way to pass the slice data into this file from another source. Likely the easiest thing to do is to list all of your unique SD/SST value combinations as literals here, restart Open5GS in your environment (docker compose down/up), and then the registration requests should stop being rejected.

```bash
slice_data = [
        {
            "sst": 1,
            "default_indicator": True,
            "session": [
                {
                    "name": apn,
                    "type": 3, "pcc_rule": [], "ambr": {"uplink": {"value": 1, "unit": 3}, "downlink": {"value": 1, "unit": 3}},
                    "qos": {
                        "index": int(qci),
                        "arp": {"priority_level": 8, "pre_emption_capability": 1, "pre_emption_vulnerability": 1}
                    },
                    "ue": {
                        "ipv4": ip_alloc
                    }
                },
                {
                    "name": "ims",
                    "type": 3, "pcc_rule": [], "ambr": {"uplink": {"value": 1, "unit": 3}, "downlink": {"value": 1, "unit": 3}},
                    "qos": {
                        "index": 5,
                        "arp": {"priority_level": 8, "pre_emption_capability": 1, "pre_emption_vulnerability": 1}
                    }
                }
            ]
        }
    ]
```

**Comment by AlejandraOliver** [comment]
Hi, @rcbarke was right!! The add_users.py needs to be configured to!!

Also, @pgawlowicz, I wanted to ask you about being able to see that each user is in a slice. As you shown, you can see that each UE has connected to a different slice using the Wireshark captures in NGAP like:

<img width="909" alt="Captura de pantalla 2025-03-13 004255" src="https://github.com/user-attachments/assets/d7869b5c-620f-484d-a1aa-b4eeb90e0dcb" />


 I've followed these instructions abou DTL_USER and NAS-5GS (https://docs.srsran.com/projects/project/en/latest/user_manuals/source/outputs.html#e2ap-pcap) to decode and view the messages in wireshark, but in the gnb_ngap.pcap I get is the following:

![image](https://github.com/user-attachments/assets/7372e0a7-66e6-41a6-a088-f5bc440630d3)

can you tell me how tomdo it @pgawlowicz? Thanks!
